waiter/src/waiter/auth/authentication.clj

```diff
@@ -26,12 +27,32 @@
   (wrap-auth-handler [this request-handler]
     "Attaches middleware that enables the application to perform authentication.
      The middleware should
-     - either issue a 401 challenge asking the client to authenticate itself,
+     - issue a 401 challenge, or redirect, to get the client to authenticate itself,
      - or upon successful authentication populate the request with :authorization/user and :authorization/principal"))
 
+(defprotocol CompositeAuthenticator
+  (get-authentication-providers [this]
+    "Get a list of supported authentication provider names."))
+
+(extend-protocol CompositeAuthenticator
+  Object
+  (get-authentication-providers [_] []))
+
+(defprotocol CallbackAuthenticator
```

Comment:
A `CallbackAuthenticator` is not an `Authenticator`, it is only listing the processing additional authentication paths. Maybe we use a different name to avoid this confusion - `OfflineAuthenticationProcessor`?